File: src/egs/misp/local/utils/create_symbol_table.py
#!/usr/bin/python3

import sys

if __name__ != '__main__':
    raise ImportError ('This script can only be run, and can\'t be imported')

if len(sys.argv) < 2:
    raise TypeError ('USAGE: create_symbol_table.py <input-symbols> [<old-table>]')

symbols_file   = sys.argv[1]
old_table_file = None
if len(sys.argv) == 3:
    old_table_file = sys.argv[2]

counter = 0
dict = {}

# Read the old table if available
if old_table_file is not None:
    with open(old_table_file, mode='r') as f:
        for line in f:
            key, val = line.split()
            dict[key] = int(val)

# Update the counter
counter = len(dict)

# Read input symbols
symbols = []
with open(symbols_file) as f:
    for line in f:
        symbols.append(line.split()[0])

# Add new symbols into the dictionary
for s in symbols:
    if s not in dict:
        dict[s] = counter
        counter = counter + 1
        
# Write the dictionary to the output
for key, val in dict.items():
    print("%s %d" % (key, val))

File: src/egs/misp/local/data/create_symbol_table.pl
#!/usr/bin/env perl
# Copyright 2018 Peter Gazdik

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
# THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
# WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
# MERCHANTABLITY OR NON-INFRINGEMENT.
# See the Apache 2 License for the specific language governing permissions and
# limitations under the License.

# Creates an FST-like symbol table from the given list of symbols or
# optionaly reuses an existing symbol table.

# Process optional command options
for ($x = 0; $x < 1; $x++) {
    if ($ARGV[0] eq "--old-symtab") {
        shift @ARGV;
        $old_symtab = shift @ARGV;
        if ($old_symtab !~ /^.+\.txt$/ || $old_symtab eq "") {
            die "The --old-symtab option requires an argument.";
        }
    }
}

# Read the existing symbol table if it's given
if (defined $old_symtab) {
    open(F, $old_symtab) || die "Error opening symbol table file $old_symtab";
    while(<F>) {
        @A = split(" ", $_);
        @A == 2 || die "Bad line in symbol table file: $_";
        $sym2int{$A[0]} = $A[1] + 0;
    }
}

# Fill the symbol table
$symbols = shift @ARGV;
if (!defined $symbols) {
    print STDERR "Usage: create_symbol_table.pl [options] <symbols> > <symbol_table>\n" .
        "options: [--old-symtab <old_symtab>] \n";
}
open(F, $symbols) || die "Error opening file $symbols.";

$idx = keys %sym2int;
while (<F>) {
    chomp($_);
    $sym2int{$_} = $idx++;
}
if ($idx == 0) {
    die "There are no symbols in input file.";
}

# Write the symbol table to the output
while (($sym, $val) = each(%sym2int)) {
    print $sym, " ", $val, "\n"
}
